src/Raven.Server/Documents/ShardedTcpHandlers/ShardedSubscriptionConnection.cs

```diff
@@ -0,0 +1,1209 @@
+        private bool _isDisposed;
+
+        private ShardedSubscriptionConnection(ServerStore serverStore, TcpConnectionOptions tcpConnectionOptions, IDisposable tcpConnectionDisposable, JsonOperationContext.MemoryBuffer buffer)
+            : base(tcpConnectionOptions, serverStore, buffer, tcpConnectionDisposable)
+        {
+        }
+
+        //TODO: egor handle double code with InitAsync and AssertConnection
+        public async Task TryReconnectRejectedShards()
+        {
+            var sp = new Stopwatch();
+            while (CancellationTokenSource.IsCancellationRequested == false)
+            {
+                if (_rejectedShards.Count == 0)
+                {
+                    await Task.Delay(1000);
+                    if(sp.IsRunning)
+                        sp.Reset();
+
+                    continue;
+                }
+
+                var rejectedShards = _rejectedShards;
+                if (CancellationTokenSource.IsCancellationRequested)
+                    break;
+
+                if (sp.IsRunning == false)
+                    sp.Start();
+      
+                if (sp.Elapsed >= _rejectedReconnectTimeout)
+                    throw new SubscriptionConnectionDownException($"Could not reconnect shards '{string.Join(",", rejectedShards)}' stopping sharded subscription '{Options.SubscriptionName}'.");
+
+                var workers = new List<SubscriptionServerWorker>();
+                var skippedNodes = new Dictionary<string, List<string>>();
+                foreach (var shard in rejectedShards)
+                {
+                    skippedNodes[shard] = new List<string>();
+                    var shardId = ShardHelper.TryGetShardIndex(shard);
+                    var re = TcpConnection.ShardedContext.RequestExecutors[shardId];
+                    if (re.Topology == null)
+                    {
+                        await re.WaitForTopologyUpdate(re._firstTopologyUpdate);
+                        if (re.Topology == null)
+                        {
+                            throw new Exception("ShouldNot happen and likely a bug.!!");
+                        }
+                    }
+
+                    SubscriptionServerWorker worker = null;
+                    List<Exception> exceptions = new List<Exception>();
+                    foreach (var node in re.Topology.Nodes)
+                    {
+                        try
+                        {
+                            worker = await GetShardedSubscriptionServerWorker(shard, re, node.ClusterTag);
+                        }
+                        catch (Exception ex)
+                        {
+                            exceptions.Add(ex);
+                            skippedNodes[shard].Add(node.ClusterTag);
+                            continue;
+                        }
+
+                        workers.Add(worker);
+                        break;
+                    }
+                }
```

Comment:
why do you need this code at all? Why not just try to connect, and let the request exec / subscription connection handle this?